arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+          consecutiveTriesWithoutProgress = 0;
+        }
+      }
+      finished = true;
+    } catch (basics::Exception const& exception) {
+      LOG_CTX("eadb8", ERR, _logContext)
+          << "Caught exception while destroying a log leader: " << exception.message();
+    } catch (std::exception const& exception) {
+      LOG_CTX("35d0b", ERR, _logContext)
+          << "Caught exception while destroying a log leader: " << exception.what();
+    } catch (...) {
+      LOG_CTX("0c972", ERR, _logContext)
+          << "Caught unknown exception while destroying a log leader!";
+    }
+    if (!finished && consecutiveTriesWithoutProgress > 10) {
+      LOG_CTX("d5d25", FATAL, _logContext)
+          << "We keep failing at destroying a log leader instance. Giving up "
+             "now.";
+      FATAL_ERROR_EXIT();
+    }
+  } while (!finished);
+}
+
+auto replicated_log::LogLeader::instantiateFollowers(
+    LoggerContext logContext, std::vector<std::shared_ptr<AbstractFollower>> const& follower,
+    std::shared_ptr<LocalFollower> const& localFollower, TermIndexPair lastEntry)
+    -> std::vector<FollowerInfo> {
+  auto initLastIndex = lastEntry.index.saturatedDecrement();
+
+  std::vector<FollowerInfo> follower_vec;
+  follower_vec.reserve(follower.size() + 1);
+  follower_vec.emplace_back(localFollower, lastEntry, logContext);
+  std::transform(follower.cbegin(), follower.cend(), std::back_inserter(follower_vec),
+                 [&](std::shared_ptr<AbstractFollower> const& impl) -> FollowerInfo {
+                   return FollowerInfo{impl, TermIndexPair{LogTerm{0}, initLastIndex}, logContext};
+                 });
+  return follower_vec;
+}
+
+namespace {
+auto delayedFuture(std::chrono::steady_clock::duration duration)
+    -> futures::Future<futures::Unit> {
+  if (SchedulerFeature::SCHEDULER) {
+    return SchedulerFeature::SCHEDULER->delay(duration);
+  }
+
+  // std::this_thread::sleep_for(duration);
+  return futures::Future<futures::Unit>{std::in_place};
+}
+}  // namespace
+
+void replicated_log::LogLeader::handleResolvedPromiseSet(
+    ResolvedPromiseSet resolvedPromises,
+    std::shared_ptr<ReplicatedLogMetrics> const& logMetrics) {
+  auto const commitTp = InMemoryLogEntry::clock::now();
+  for (auto const& it : resolvedPromises._commitedLogEntries) {
+    using namespace std::chrono_literals;
+    auto const entryDuration = commitTp - it.insertTp();
+    logMetrics->replicatedLogInsertsRtt->count(entryDuration / 1us);
+  }
+
+  for (auto& promise : resolvedPromises._set) {
+    TRI_ASSERT(promise.second.valid());
+    promise.second.setValue(resolvedPromises._quorum);
+  }
+}
+
+void replicated_log::LogLeader::executeAppendEntriesRequests(
+    std::vector<std::optional<PreparedAppendEntryRequest>> requests,
+    std::shared_ptr<ReplicatedLogMetrics> const& logMetrics) {
+  for (auto& it : requests) {
+    if (it.has_value()) {
+      delayedFuture(it->_executionDelay).thenFinal([it = it, logMetrics](auto&&) mutable {
```

Comment:
This seems to copy `it` for every request. Can it be moved instead?

Reply:
I moved `it` into the lambda.